src/event/workqueue.c

```diff
@@ -0,0 +1,287 @@
+/* Are we using user-level or kext based management? */
+static bool _dispatch_workq_kext_active;
+#endif
+
+/*
+ * State for the user-level monitoring of a workqueue.
+ */
+typedef struct dispatch_workq_manager_s {
+	/* The observed number of runnable registered workers */
+	int32_t runnable_workers;
+
+	/* The desired number of runnable registered workers */
+	int32_t target_runnable_workers;
+
+	/*
+     * Tracking of registered workers; all accesses must hold lock.
+     * Invariant: registered_workers[0]...registered_workers[num_registered_workers-1]
+     *   contain the pids of the workers that we are managing.
+     */
+	dispatch_unfair_lock_s registered_worker_lock;
+	pid_t *registered_workers;
+	int num_registered_workers;
+} dispatch_workq_manager_s, *dispatch_workq_manager_t;
+
+static dispatch_workq_manager_s _dispatch_workq_manager;
+
+#pragma mark inline helper functions
+
+DISPATCH_INLINE
+dispatch_workq_manager_t
+_dispatch_workq_get_default_manager() {
+	return &_dispatch_workq_manager;
+}
+
+DISPATCH_INLINE
+dispatch_queue_t
+_dispatch_workq_get_default_root_queue()
+{
+	return &_dispatch_root_queues[DISPATCH_ROOT_QUEUE_IDX_DEFAULT_QOS];
+}
+
+DISPATCH_INLINE
+bool
+_dispatch_workq_root_queue_has_work(dispatch_queue_t dq)
+{
+	return dq->dq_items_tail != NULL;
+}
+
+#pragma mark Implementation of the management subsystem.
+
+#define WORKQ_MAX_TRACKED_WORKERS DISPATCH_WORKQ_MAX_PTHREAD_COUNT
+#define WORKQ_OVERSUBSCRIBE_FACTOR 2
+
+static void _dispatch_workq_init_once(void *context DISPATCH_UNUSED);
+static dispatch_once_t _dispatch_workq_init_once_pred;
+
+bool
+dispatch_workq_worker_register(dispatch_queue_t root_q)
+{
+	if (root_q != _dispatch_workq_get_default_root_queue()) {
+		// Not tracked.
+		return false;
+	}
+
+	dispatch_once_f(&_dispatch_workq_init_once_pred, NULL, &_dispatch_workq_init_once);
+
+#if DISPATCH_ENABLE_PWQ_KEXT
+	if (_dispatch_workq_kext_active) {
+		_dispatch_workq_worker_register_kext();
+		return true;
+	}
+#endif
+	dispatch_workq_manager_t mgr = _dispatch_workq_get_default_manager();
+	bool rc;
+	int tid = syscall(SYS_gettid);
+	_dispatch_unfair_lock_lock(&mgr->registered_worker_lock);
+	if (mgr->num_registered_workers < WORKQ_MAX_TRACKED_WORKERS-1) {
+		int worker_id = mgr->num_registered_workers++;
+		mgr->registered_workers[worker_id] = tid;
+		rc = true;
+	} else {
+		rc = false;
+	}
+	_dispatch_unfair_lock_unlock(&mgr->registered_worker_lock);
+
+	return rc;
+}
+
+void
+dispatch_workq_worker_unregister(dispatch_queue_t root_q)
+{
+	if (root_q != _dispatch_workq_get_default_root_queue()) {
+		// Not tracked.
+		return;
+	}
+
+#if DISPATCH_ENABLE_PWQ_KEXT
+	if (_dispatch_workq_kext_active) {
+		_dispatch_workq_worker_register_kext();
+		return;
+	}
+#endif
+	dispatch_workq_manager_t mgr = _dispatch_workq_get_default_manager();
+	int tid = syscall(SYS_gettid);
+	_dispatch_unfair_lock_lock(&mgr->registered_worker_lock);
+	for (int i=0; i<mgr->num_registered_workers; i++) {
```

Comment:
It is even more critical that you will likely when adding feature to the pool want access to others of these TSDs anyway.